tests/test_class_sh_basic.cpp

```diff
@@ -64,6 +64,28 @@ std::unique_ptr<atyp const, sddc> rtrn_udcp() { return std::unique_ptr<atyp cons
 std::string pass_udmp(std::unique_ptr<atyp,       sddm> obj) { return "pass_udmp:" + obj->mtxt; }
 std::string pass_udcp(std::unique_ptr<atyp const, sddc> obj) { return "pass_udcp:" + obj->mtxt; }
 
+
+struct custom_deleter
+{
+    std::string delete_txt;
+    custom_deleter() = default;
```

Comment:
I think I can resolve the problems around the actual deleter, but this change also trips over this line:

https://github.com/pybind/pybind11/blob/edfaaed87ea3a869eaa1947f3d40cb62958295ca/include/pybind11/detail/smart_holder_type_casters.h#L592

Since that requires the deleter to be default constructed. Is it possible to encounter that situation and does it require a `nullptr` or can it be handled with a throw?

Reply:
I don't remember TBH. Or it's simply something I swept under the rug during the smart_holder bootstrap phase and I never ran into a problem with it.

Brute force idea: I could first run local testing, then maybe global testing, to see if/how that path is used.

How troublesome is this to implement: Return `nullptr` if `D` is default-constructible, or throw an exception otherwise?

But also: if we get into more trouble like this, I'm OK to not solve this for now.

Reply:
> How troublesome is this to implement: Return nullptr if D is default-constructible, or throw an exception otherwise?

Yeah, we need a sfinea function anyway to do this, since;


```cpp
std::unique_ptr<T, D> result;
if (extracted_deleter != nullptr) {
   result = std::unique_ptr<T, D>(raw_type_ptr, std::move(*extracted_deleter);
} else {
   result = std::unique_ptr<T, D>(raw_type_ptr);

}
```

would require `D` to be default constructible...  So we need to branch at compile time on whether `D` is default constructible anyway. I think I got something that works, let me do a quick format and let me push that commit such that you can see what complexity it brings.

Reply:
See 62bb88ed57c0ae9fb1109ec5459b7feb6980d287

It's doable, but like I expected it does add complexity, we can also remove the commit, your call. :)